### PR TITLE
added clean return data endpoint

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

File: arwen/contexts/output.go [hunk 1]
```diff
@@ -201,6 +201,14 @@ func (context *outputContext) ClearReturnData() {
 	context.outputState.ReturnData = make([][]byte, 0)
 }
 
+func (context *outputContext) RemoveReturnData(index uint32) {
```
Comment (sasurobert) [inline arwen/contexts/output.go:204]: add comment for public function

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -3222,24 +3236,70 @@ func v1_4_getReturnDataSize(context unsafe.Pointer, resultID int32) int32 {
...
+	output.ClearReturnData()
+}
+
+//export v1_4_deleteFromReturnData
+func v1_4_deleteFromReturnData(context unsafe.Pointer, resultID int32) {
+	host := arwen.GetVMHost(context)
+	DeleteFromReturnDataWithHost(host, resultID)
+}
+
+func DeleteFromReturnDataWithHost(host arwen.VMHost, resultID int32) {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:3292]: comment on public function

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -3222,24 +3236,70 @@ func v1_4_getReturnDataSize(context unsafe.Pointer, resultID int32) int32 {
...
+}
 
-	return int32(len(returnData[resultID]))
+//export v1_4_cleanReturnData
+func v1_4_cleanReturnData(context unsafe.Pointer) {
+	host := arwen.GetVMHost(context)
+	CleanReturnDataWithHost(host)
+}
+
+func CleanReturnDataWithHost(host arwen.VMHost) {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:3276]: comment on public function

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -3222,24 +3236,70 @@ func v1_4_getReturnDataSize(context unsafe.Pointer, resultID int32) int32 {
...
 	if resultID >= int32(len(returnData)) {
-		return 0
+		return nil
 	}
 
-	err := runtime.MemStore(dataOffset, returnData[resultID])
-	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
-		return 0
-	}
+	return returnData[resultID]
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:3267]: `resultID` can be negative here. There should be a check that `resultID` is ≥ 0, or at least cast it to an `uint32`. Contracts written in C can pass -4 here, causing a panic.

File: arwen/contexts/output.go [hunk 5]
```diff
@@ -201,6 +201,14 @@ func (context *outputContext) ClearReturnData() {
...
+	returnData := context.outputState.ReturnData
+	if index >= uint32(len(returnData)) {
+		return
+	}
+	context.outputState.ReturnData = append(returnData[:index], returnData[index+1:]...)
```
Comment (camilbancioiu) [inline arwen/contexts/output.go:209]: Would it make sense to call `ClearReturnData` if `index` is 0 and `len(returnData) == 1`? What happens here if we remove the last element of `returnData`? Hopefully, `append` will return an empty slice, and won't return `nil`.

File: arwen/hosttest/execution_test.go [hunk 6]
```diff
@@ -2895,6 +2897,75 @@ func TestExecution_Mocked_Warm_Instances_Same_Contract(t *testing.T) {
...
+						childInput.GasProvided = 1000
+						returnValue := contracts.ExecuteOnDestContextInMockContracts(host, childInput)
+
+						require.Equal(t, int32(0), returnValue)
+						returnData := elrondapi.GetReturnDataWithHostAndTypedArgs(host, 0)
+						require.Equal(t, zero, string(returnData))
+						returnData = elrondapi.GetReturnDataWithHostAndTypedArgs(host, 1)
+						require.Equal(t, one, string(returnData))
+						returnData = elrondapi.GetReturnDataWithHostAndTypedArgs(host, 2)
+						require.Equal(t, two, string(returnData))
```
Comment (camilbancioiu) [inline arwen/hosttest/execution_test.go:2925]: Consider adding cases with negative indices.

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

File: arwen/contexts/output.go (hunk 5 above)
Comment (matei-p) [reply arwen/contexts/output.go:210]: Your hopes are true :) A test was just added.

**Review by sasurobert — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**